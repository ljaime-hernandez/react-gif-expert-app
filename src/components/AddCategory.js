import React, {useState} from 'react'
import PropTypes from 'prop-types';

// the props 'setCategories' are destructured before being used on the component, done that we can call the 
// setCategories function from the parent component to properly add the input value into the string array declared
// on it
export const AddCategory = ({setCategories}) => {
    
    // the inputValue by default wll be left as an empty string on the useState, if we remove the 
    // quotation marks then the browser will send an error, the error will not be explicit but it will be
    // a reference to the inputValue object stating that it is in an undefined state, when the 
    // quotation marks are placed on the useState then the error will disappear
    const [inputValue, setInputValue] = useState('');

    // this function will retrieve the value placed in the input box, done by calling the event, 
    // then the target value will contain the input value which will be assigned to the inputValue
    // object which we will use to include into the categories from the previous component
    const handleInputChange = (e) => {
        setInputValue(e.target.value);
    }

    // we use the handleSubmit function to first prevent the input box from having an unrequired submission
    // with an incorrect input, then we check if the input value is greater than 2 characters, if its true
    // then the prop setCategories from the previous component is called for it to include the string input in
    // the categories object
    const handleSubmit = (e) => {
        e.preventDefault();

        if (inputValue.trim().length > 2){
            // inputvalue is placed first for it to push the other results to the bottom on the screen and have
            // the new results for quick reading
            setCategories(category => [inputValue, ...category]);
            // sets the text input box to an empty state
            setInputValue('');
        }
        
        console.log('Submit succeed')
    }

    return (
        <form onSubmit={handleSubmit}>
            {/*'p' element added just for testing purposes, more information on the file located in
            tests/components/AddCategory.test.js file*/}
            <p hidden>{inputValue}</p>
            <input
                type="text"
                value={inputValue}
                onChange={handleInputChange}
            />
        </form>
    )
}

// the PropTypes will send an error on the browser console if the AddCategory component is used
// without the setCategories sent as an argument, this will add a security layer which will allow
// to only make modifications to a get or post request under our rules
AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}