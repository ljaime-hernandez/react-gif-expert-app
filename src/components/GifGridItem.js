import React from 'react';

export const GifGridItem = ({title, url}) => {

    return (
        <div className="card animate__animated animate__fadeInLeft">
            <img src={url} alt={title}/>
            <p>{title}</p>
        </div>
    )
}
 