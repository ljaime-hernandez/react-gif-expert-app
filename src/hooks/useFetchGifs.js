import { useState, useEffect } from "react";
import { getGifs } from "../helpers/getGifs";

// custom hook created for it to load our API request with a custom loading state
// the data will be an empty array with a loading state of 'true', when the setState 
// function will be used along with a useEffect which will be loaded every time a new 
// parameter is placed in the search bar, the component will then be loaded again with 
// new categories, which will be rendered again and the loading state will be false,
// the setState is contained on a timeout function which will update itself after 3 seconds

export const useFetchGifs = (category) => {

    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect(() => {
        
        getGifs(category)
            .then(imgs => {
                setTimeout(() => {
                    setState({
                        data: imgs,
                        loading: false
                    })
                }, 1000)
            })
    }, [category])

    return state;
}
