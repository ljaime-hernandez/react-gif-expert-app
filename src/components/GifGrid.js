import React from 'react'
// import { getGifs } from '../helpers/getGifs';
import { GifGridItem } from './GifGridItem';
import { useFetchGifs } from '../hooks/useFetchGifs';
import PropTypes from 'prop-types';

export const GifGrid = ({category}) => {

    // const [images, setImages] = useState([]);

    const {data:images, loading} = useFetchGifs(category);

    // // function run every time the component is loaded
    // useEffect(() => {
    //     getGifs(category)
    //         .then(imgs => setImages(imgs));
    //         // also .then(setImages);

    // // in determined cases, the argument sent from the parent component might change instead of adding an additional
    // // one, a warning will be sent through the console for us to acknowledge a lack of cohesion on the useEffect, in
    // // this code we just need to add the category argument as a dependency inside the array at the end of the function     
    // }, [category])

    return (
        <>
            <h3>{category}</h3>

            {loading && <p>Loading</p>}
            {/* the operator && will work for us as an evaluation of a true/false statement, it can be used
            instead of a ternary evaluation, the ternary option is placed below as example for a different approach.
            {loading ? <p>Loading</p> : null}
            */}

            {/*we use className instead of class as javascript will give us a warning when rendering the tag, as "class"
            is a reserved word for scripting uses*/}
             <div className="card-grid">
                {
                    
                    images.map( image => (
                            <GifGridItem 
                                key={image.id}
                                {...image} />))
                }
            </div> 
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired,
}