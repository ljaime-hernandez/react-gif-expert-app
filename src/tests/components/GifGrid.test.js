import React from "react";
import { shallow } from "enzyme";
import { GifGrid } from "../../components/GifGrid";
import { useFetchGifs } from "../../hooks/useFetchGifs";

// we use the jest library to create a mock return from the function used as
// hook imported from useFetchGifs.js
jest.mock("../../hooks/useFetchGifs");

describe('Tests on <GifGrid />', () =>{

    const category = 'Gantz';

    test('should display component correctly', () => {
        
        // we use the mockReturnValue to return the object containing an array
        // and a boolean used for the GifGrid component to work as the hook 
        // used on it requires the useFetchGifs to be called at least once
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true,
        });
        const wrapper = shallow(<GifGrid category={category}/>);   
        expect(wrapper).toMatchSnapshot();
    });

    test('should show items when loading images from useFetchGifs', () => {
    
        // object created as mock to be added in the following function mock
        const gifs = [{
            id: 'Gantz',
            url: 'https://gantz.org',
            title: 'Gantz Title',
        }]
        // simulation of a return of the useFetchGifs hook return value to be
        // used on the snapshot
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false,
        });
        // the p element should not be displayed as its only shown when theres no
        // response from the useFetchGifs hook 
        const wrapper = shallow(<GifGrid category={category}/>)
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('p').exists()).toBe(false);
        // the GifGridItem component should have been rendered the number of times 
        // the useFetchGifs returned objects, in this case the number of objects is
        // just one
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length);
    });
    
})