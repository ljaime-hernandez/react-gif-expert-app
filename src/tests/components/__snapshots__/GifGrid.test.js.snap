// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <GifGrid /> should display component correctly 1`] = `
<Fragment>
  <h3>
    Gantz
  </h3>
  <p>
    Loading
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Tests on <GifGrid /> should show items when loading images from useFetchGifs 1`] = `
<Fragment>
  <h3>
    Gantz
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="Gantz"
      key="Gantz"
      title="Gantz Title"
      url="https://gantz.org"
    />
  </div>
</Fragment>
`;
